File: ios/VisionUtilsTypes.swift
import Foundation
import UIKit
import Photos
import Accelerate

// MARK: - Option Structs (mirrored from TypeScript)

/// Image source types
enum ImageSourceType: String {
    case url
    case file
    case base64
    case asset
    case photoLibrary
}

/// Color format options
enum ColorFormat: String {
    case rgb
    case rgba
    case bgr
    case bgra
    case grayscale
    case hsv
    case hsl
    case lab
    case yuv
    case ycbcr

    var channelCount: Int {
        switch self {
        case .rgb, .bgr, .hsv, .hsl, .lab, .yuv, .ycbcr: return 3
        case .rgba, .bgra: return 4
        case .grayscale: return 1
        }
    }
}

/// Resize strategy options
enum ResizeStrategy: String {
    case cover
    case contain
    case stretch
    case letterbox
}

/// Data layout options
enum DataLayout: String {
    case hwc
    case chw
    case nhwc
    case nchw

    var hasBatchDimension: Bool {
        return self == .nhwc || self == .nchw
    }

    var isChannelFirst: Bool {
        return self == .chw || self == .nchw
    }
}

/// Normalization preset types
enum NormalizationPreset: String {
    case imagenet
    case tensorflow
    case scale
    case raw
    case custom
}

/// Image source configuration
struct ImageSource {
    let type: ImageSourceType
    let value: String

    init(from dict: [String: Any]) throws {
        guard let typeStr = dict["type"] as? String,
              let type = ImageSourceType(rawValue: typeStr) else {
            throw VisionUtilsError.invalidSource("Invalid or missing source type")
        }

        guard let value = dict["value"] as? String else {
            throw VisionUtilsError.invalidSource("Missing source value")
        }

        self.type = type
        self.value = value
    }
}

/// Resize options
struct ResizeOptions {
    let width: Int
    let height: Int
    let strategy: ResizeStrategy
    let padColor: [UInt8]

    init(from dict: [String: Any]) throws {
        guard let width = dict["width"] as? Int,
              let height = dict["height"] as? Int else {
            throw VisionUtilsError.invalidResize("Width and height are required")
        }

        self.width = width
        self.height = height

        if let strategyStr = dict["strategy"] as? String,
           let strategy = ResizeStrategy(rawValue: strategyStr) {
            self.strategy = strategy
        } else {
            self.strategy = .cover
        }

        if let padColorArr = dict["padColor"] as? [Int] {
            self.padColor = padColorArr.map { UInt8(clamping: $0) }
        } else {
            self.padColor = [0, 0, 0, 255]
        }
    }
}

/// Region of interest
struct Roi {
    let x: Int
    let y: Int
    let width: Int
    let height: Int

    init(from dict: [String: Any]) throws {
        guard let x = dict["x"] as? Int,
              let y = dict["y"] as? Int,
              let width = dict["width"] as? Int,
              let height = dict["height"] as? Int else {
            throw VisionUtilsError.invalidRoi("ROI requires x, y, width, and height")
        }

        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

/// Normalization configuration
struct Normalization {
    let preset: NormalizationPreset
    let mean: [Float]?
    let std: [Float]?
    let scale: Float

    init(from dict: [String: Any]) throws {
        guard let presetStr = dict["preset"] as? String,
              let preset = NormalizationPreset(rawValue: presetStr) else {
            throw VisionUtilsError.invalidNormalization("Invalid normalization preset")
        }

        self.preset = preset

        if preset == .custom {
            guard let mean = dict["mean"] as? [Double],
                  let std = dict["std"] as? [Double] else {
                throw VisionUtilsError.invalidNormalization("Custom normalization requires mean and std arrays")
            }
            self.mean = mean.map { Float($0) }
            self.std = std.map { Float($0) }
        } else {
            self.mean = nil
            self.std = nil
        }

        if let scale = dict["scale"] as? Double {
            self.scale = Float(scale)
        } else {
            self.scale = 1.0 / 255.0
        }
    }

    static let `default` = Normalization(preset: .raw, mean: nil, std: nil, scale: 1.0)

    private init(preset: NormalizationPreset, mean: [Float]?, std: [Float]?, scale: Float) {
        self.preset = preset
        self.mean = mean
        self.std = std
        self.scale = scale
    }
}

/// Complete options for getPixelData
struct GetPixelDataOptions {
    let source: ImageSource
    let colorFormat: ColorFormat
    let normalization: Normalization
    let resize: ResizeOptions?
    let roi: Roi?
    let layout: DataLayout

    init(from dict: [String: Any]) throws {
        guard let sourceDict = dict["source"] as? [String: Any] else {
            throw VisionUtilsError.invalidSource("Source is required")
        }

        self.source = try ImageSource(from: sourceDict)

        if let colorFormatStr = dict["colorFormat"] as? String,
           let colorFormat = ColorFormat(rawValue: colorFormatStr) {
            self.colorFormat = colorFormat
        } else {
            self.colorFormat = .rgb
        }

        if let normDict = dict["normalization"] as? [String: Any] {
            self.normalization = try Normalization(from: normDict)
        } else {
            self.normalization = Normalization.default
        }

        if let resizeDict = dict["resize"] as? [String: Any] {
            self.resize = try ResizeOptions(from: resizeDict)
        } else {
            self.resize = nil
        }

        if let roiDict = dict["roi"] as? [String: Any] {
            self.roi = try Roi(from: roiDict)
        } else {
            self.roi = nil
        }

        if let layoutStr = dict["dataLayout"] as? String,
           let layout = DataLayout(rawValue: layoutStr.lowercased()) {
            self.layout = layout
        } else {
            self.layout = .hwc
        }
    }
}

// MARK: - Error Types

enum VisionUtilsError: Error {
    case invalidSource(String)
    case loadError(String)
    case fileNotFound(String)
    case permissionDenied(String)
    case processingError(String)
    case invalidRoi(String)
    case invalidResize(String)
    case invalidNormalization(String)
    case unknown(String)

    var code: String {
        switch self {
        case .invalidSource: return "INVALID_SOURCE"
        case .loadError: return "LOAD_ERROR"
        case .fileNotFound: return "FILE_NOT_FOUND"
        case .permissionDenied: return "PERMISSION_DENIED"
        case .processingError: return "PROCESSING_ERROR"
        case .invalidRoi: return "INVALID_ROI"
        case .invalidResize: return "INVALID_RESIZE"
        case .invalidNormalization: return "INVALID_NORMALIZATION"
        case .unknown: return "UNKNOWN"
        }
    }

    var message: String {
        switch self {
        case .invalidSource(let msg),
             .loadError(let msg),
             .fileNotFound(let msg),
             .permissionDenied(let msg),
             .processingError(let msg),
             .invalidRoi(let msg),
             .invalidResize(let msg),
             .invalidNormalization(let msg),
             .unknown(let msg):
            return msg
        }
    }
}

// MARK: - Result Type

struct VisionUtilsResult {
    let data: [Float]
    let width: Int
    let height: Int
    let channels: Int
    let colorFormat: ColorFormat
    let layout: DataLayout
    let shape: [Int]
    let processingTimeMs: Double

    func toDictionary() -> [String: Any] {
        return [
            "data": data,
            "width": width,
            "height": height,
            "channels": channels,
            "colorFormat": colorFormat.rawValue,
            "dataLayout": layout.rawValue.uppercased(),
            "shape": shape,
            "processingTimeMs": processingTimeMs
        ]
    }
}

File: android/src/main/java/com/visionutils/VisionUtilsTypes.kt
package com.visionutils

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.Arguments

/**
 * Image source types supported by the library
 */
enum class ImageSourceType {
    URL,
    FILE,
    BASE64,
    ASSET,
    PHOTO_LIBRARY;

    companion object {
        fun fromString(value: String): ImageSourceType {
            return when (value.lowercase()) {
                "url" -> URL
                "file" -> FILE
                "base64" -> BASE64
                "asset" -> ASSET
                "photolibrary", "photo_library" -> PHOTO_LIBRARY
                else -> throw VisionUtilsException("INVALID_SOURCE", "Unknown image source type: $value")
            }
        }
    }
}

/**
 * Image source with type and value
 */
data class ImageSource(
    val type: ImageSourceType,
    val value: String
) {
    companion object {
        fun fromMap(map: ReadableMap): ImageSource {
            val type = map.getString("type")
                ?: throw VisionUtilsException("INVALID_SOURCE", "Missing source type")
            val value = map.getString("value")
                ?: throw VisionUtilsException("INVALID_SOURCE", "Missing source value")
            return ImageSource(ImageSourceType.fromString(type), value)
        }
    }
}

/**
 * Color format for pixel data output
 */
enum class ColorFormat {
    RGB,
    RGBA,
    BGR,
    BGRA,
    GRAYSCALE,
    HSV,
    HSL,
    LAB,
    YUV,
    YCBCR;

    companion object {
        fun fromString(value: String): ColorFormat {
            return when (value.uppercase()) {
                "RGB" -> RGB
                "RGBA" -> RGBA
                "BGR" -> BGR
                "BGRA" -> BGRA
                "GRAYSCALE" -> GRAYSCALE
                "HSV" -> HSV
                "HSL" -> HSL
                "LAB" -> LAB
                "YUV" -> YUV
                "YCBCR" -> YCBCR
                else -> RGB // Default
            }
        }
    }

    val channels: Int
        get() = when (this) {
            RGB, BGR, HSV, HSL, LAB, YUV, YCBCR -> 3
            RGBA, BGRA -> 4
            GRAYSCALE -> 1
        }
}

/**
 * Resize strategy
 */
enum class ResizeStrategy {
    COVER,
    CONTAIN,
    STRETCH,
    LETTERBOX;

    companion object {
        fun fromString(value: String): ResizeStrategy {
            return when (value.lowercase()) {
                "cover" -> COVER
                "contain" -> CONTAIN
                "stretch" -> STRETCH
                "letterbox" -> LETTERBOX
                else -> COVER // Default
            }
        }
    }
}

/**
 * Resize options
 */
data class ResizeOptions(
    val width: Int,
    val height: Int,
    val strategy: ResizeStrategy = ResizeStrategy.COVER,
    val padColor: IntArray = intArrayOf(0, 0, 0, 255),
    val letterboxColor: IntArray = intArrayOf(114, 114, 114)
) {
    companion object {
        fun fromMap(map: ReadableMap?): ResizeOptions? {
            if (map == null) return null
            val width = if (map.hasKey("width")) map.getInt("width") else return null
            val height = if (map.hasKey("height")) map.getInt("height") else return null
            val strategy = if (map.hasKey("strategy")) {
                ResizeStrategy.fromString(map.getString("strategy") ?: "cover")
            } else {
                ResizeStrategy.COVER
            }
            val padColor = if (map.hasKey("padColor")) {
                map.getArray("padColor")?.toIntArray() ?: intArrayOf(0, 0, 0, 255)
            } else {
                intArrayOf(0, 0, 0, 255)
            }
            val letterboxColor = if (map.hasKey("letterboxColor")) {
                map.getArray("letterboxColor")?.toIntArray() ?: intArrayOf(114, 114, 114)
            } else {
                intArrayOf(114, 114, 114)
            }
            return ResizeOptions(width, height, strategy, padColor, letterboxColor)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as ResizeOptions
        return width == other.width &&
               height == other.height &&
               strategy == other.strategy &&
               padColor.contentEquals(other.padColor) &&
               letterboxColor.contentEquals(other.letterboxColor)
    }

    override fun hashCode(): Int {
        var result = width
        result = 31 * result + height
        result = 31 * result + strategy.hashCode()
        result = 31 * result + padColor.contentHashCode()
        result = 31 * result + letterboxColor.contentHashCode()
        return result
    }
}

/**
 * Region of interest for cropping
 */
data class Roi(
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int
) {
    companion object {
        fun fromMap(map: ReadableMap?): Roi? {
            if (map == null) return null
            val x = if (map.hasKey("x")) map.getInt("x") else 0
            val y = if (map.hasKey("y")) map.getInt("y") else 0
            val width = if (map.hasKey("width")) map.getInt("width") else return null
            val height = if (map.hasKey("height")) map.getInt("height") else return null
            return Roi(x, y, width, height)
        }
    }
}

/**
 * Normalization presets
 */
enum class NormalizationPreset {
    IMAGENET,
    TENSORFLOW,
    SCALE,
    RAW,
    CUSTOM;

    companion object {
        fun fromString(value: String): NormalizationPreset {
            return when (value.lowercase()) {
                "imagenet" -> IMAGENET
                "tensorflow" -> TENSORFLOW
                "scale" -> SCALE
                "raw" -> RAW
                "custom" -> CUSTOM
                else -> SCALE
            }
        }
    }
}

/**
 * Normalization configuration
 */
data class Normalization(
    val preset: NormalizationPreset = NormalizationPreset.SCALE,
    val mean: FloatArray = floatArrayOf(0f, 0f, 0f),
    val std: FloatArray = floatArrayOf(1f, 1f, 1f)
) {
    companion object {
        fun fromMap(map: ReadableMap?): Normalization {
            if (map == null) return Normalization()

            val preset = if (map.hasKey("preset")) {
                NormalizationPreset.fromString(map.getString("preset") ?: "scale")
            } else {
                NormalizationPreset.SCALE
            }

            val mean = if (map.hasKey("mean")) {
                map.getArray("mean")?.toFloatArray() ?: floatArrayOf(0f, 0f, 0f)
            } else {
                floatArrayOf(0f, 0f, 0f)
            }

            val std = if (map.hasKey("std")) {
                map.getArray("std")?.toFloatArray() ?: floatArrayOf(1f, 1f, 1f)
            } else {
                floatArrayOf(1f, 1f, 1f)
            }

            return Normalization(preset, mean, std)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Normalization
        return preset == other.preset &&
               mean.contentEquals(other.mean) &&
               std.contentEquals(other.std)
    }

    override fun hashCode(): Int {
        var result = preset.hashCode()
        result = 31 * result + mean.contentHashCode()
        result = 31 * result + std.contentHashCode()
        return result
    }
}

/**
 * Data layout format
 */
enum class DataLayout {
    HWC,  // Height x Width x Channels (standard)
    CHW,  // Channels x Height x Width (PyTorch)
    NHWC, // Batch x Height x Width x Channels (TensorFlow)
    NCHW; // Batch x Channels x Height x Width

    companion object {
        fun fromString(value: String): DataLayout {
            return when (value.uppercase()) {
                "HWC" -> HWC
                "CHW" -> CHW
                "NHWC" -> NHWC
                "NCHW" -> NCHW
                else -> HWC
            }
        }
    }
}

/**
 * Output format type
 */
enum class OutputFormat {
    ARRAY,
    FLOAT32_ARRAY,
    UINT8_ARRAY;

    companion object {
        fun fromString(value: String): OutputFormat {
            return when (value.lowercase()) {
                "array" -> ARRAY
                "float32array", "float32_array" -> FLOAT32_ARRAY
                "uint8array", "uint8_array" -> UINT8_ARRAY
                else -> ARRAY
            }
        }
    }
}

/**
 * Complete options for getPixelData
 */
data class GetPixelDataOptions(
    val source: ImageSource,
    val colorFormat: ColorFormat = ColorFormat.RGB,
    val resize: ResizeOptions? = null,
    val roi: Roi? = null,
    val normalization: Normalization = Normalization(),
    val dataLayout: DataLayout = DataLayout.HWC,
    val outputFormat: OutputFormat = OutputFormat.ARRAY
) {
    companion object {
        fun fromMap(map: ReadableMap): GetPixelDataOptions {
            val source = if (map.hasKey("source")) {
                ImageSource.fromMap(map.getMap("source")!!)
            } else {
                throw VisionUtilsException("INVALID_SOURCE", "Missing source in options")
            }

            val colorFormat = if (map.hasKey("colorFormat")) {
                ColorFormat.fromString(map.getString("colorFormat") ?: "RGB")
            } else {
                ColorFormat.RGB
            }

            val resize = if (map.hasKey("resize")) {
                ResizeOptions.fromMap(map.getMap("resize"))
            } else {
                null
            }

            val roi = if (map.hasKey("roi")) {
                Roi.fromMap(map.getMap("roi"))
            } else {
                null
            }

            val normalization = if (map.hasKey("normalization")) {
                Normalization.fromMap(map.getMap("normalization"))
            } else {
                Normalization()
            }

            val dataLayout = if (map.hasKey("dataLayout")) {
                DataLayout.fromString(map.getString("dataLayout") ?: "HWC")
            } else {
                DataLayout.HWC
            }

            val outputFormat = if (map.hasKey("outputFormat")) {
                OutputFormat.fromString(map.getString("outputFormat") ?: "array")
            } else {
                OutputFormat.ARRAY
            }

            return GetPixelDataOptions(
                source = source,
                colorFormat = colorFormat,
                resize = resize,
                roi = roi,
                normalization = normalization,
                dataLayout = dataLayout,
                outputFormat = outputFormat
            )
        }
    }
}

/**
 * Batch processing options
 */
data class BatchOptions(
    val concurrency: Int = 4
) {
    companion object {
        fun fromMap(map: ReadableMap?): BatchOptions {
            if (map == null) return BatchOptions()
            val concurrency = if (map.hasKey("concurrency")) {
                map.getInt("concurrency")
            } else {
                4
            }
            return BatchOptions(concurrency)
        }
    }
}

/**
 * Result of pixel data extraction
 */
data class VisionUtilsResult(
    val data: FloatArray,
    val width: Int,
    val height: Int,
    val channels: Int,
    val colorFormat: ColorFormat,
    val dataLayout: DataLayout,
    val shape: IntArray,
    val processingTimeMs: Double
) {
    fun toWritableMap(): WritableMap {
        val map = Arguments.createMap()
        val dataArray = Arguments.createArray()
        data.forEach { dataArray.pushDouble(it.toDouble()) }
        map.putArray("data", dataArray)
        map.putInt("width", width)
        map.putInt("height", height)
        map.putInt("channels", channels)
        map.putString("colorFormat", colorFormat.name.lowercase())
        map.putString("dataLayout", dataLayout.name.lowercase())
        val shapeArray = Arguments.createArray()
        shape.forEach { shapeArray.pushInt(it) }
        map.putArray("shape", shapeArray)
        map.putDouble("processingTimeMs", processingTimeMs)
        return map
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as VisionUtilsResult
        return data.contentEquals(other.data) &&
               width == other.width &&
               height == other.height &&
               channels == other.channels &&
               colorFormat == other.colorFormat &&
               dataLayout == other.dataLayout &&
               shape.contentEquals(other.shape)
    }

    override fun hashCode(): Int {
        var result = data.contentHashCode()
        result = 31 * result + width
        result = 31 * result + height
        result = 31 * result + channels
        result = 31 * result + colorFormat.hashCode()
        result = 31 * result + dataLayout.hashCode()
        result = 31 * result + shape.contentHashCode()
        return result
    }
}

/**
 * Custom exception for vision utils errors
 */
class VisionUtilsException(
    val code: String,
    override val message: String
) : Exception(message)

// Extension functions

fun ReadableArray.toFloatArray(): FloatArray {
    val result = FloatArray(size())
    for (i in 0 until size()) {
        result[i] = getDouble(i).toFloat()
    }
    return result
}

fun ReadableArray.toIntArray(): IntArray {
    val result = IntArray(size())
    for (i in 0 until size()) {
        result[i] = getInt(i)
    }
    return result
}
